First hack and slash pass on build system/layering

benvanik opened the pull request:
![image](https://i.giphy.com/media/9tT0s3SBnDm4o/source.gif)

It's time we clean things up a bit. In preparation for packaging work I realized just how sloppy some things had gotten. This tries to clean up, document, normalize, or entirely remove where possible things standing in the way of a much smaller build configuration.

Goals are to:
* Reduce our internal and external dependencies that we need to figure out how to build/multiversion/deploy
* Cleanly separate parts of the public and private code such that we can start producing include+lib deployments that expose only the C API
* Remove unused functionality (that was either added speculatively and never used or is no longer used)
* Untangle the google3-esque build files that have variables, inclusions, etc that are impossible to reason about in cmake deployments
* Reduce the total number of special snowflake DO NOT EDIT files so bazel-to-cmake can do more (and we humans can do less)
* Reduce copt/linkopt overrides (as each adds a tax to cross-platform and cross-compilation deployments)
* For copt-style things we do need move them into code (where the build system is not involved and they are easier to find/inspect/change)
* Remove IREE_COMMON_INCLUDE_DIRS by making includes local/fixing paths/etc so that we can take deps on out-of-tree code and so that our deployments can properly pull in user-provided packages
* Remove PLATFORM_VULKAN_DEPS & co and put an actual nice place to add PLATFORM_VULKAN_DEPS-esque stuff (that is only used in google3)
* Remove alwayslink usage entirely for tblgen/HAL driver registration/etc so that we can avoid dynamic linking issues and inflicting that on downstream consumers
* Remove third_party/flatbuffers dep by switching entirely to third_party/flatcc so we only have one flatbuffers related dep
* Prepare for external usage by making it cleaner to take a dep on IREE and pull in the HAL parts
* Prepare for out-of-tree HAL backends and VM modules by removing our tight coupling on build configuration-time variables
* Improve compile/link times and build noise by reducing total target count - LTO is pretty badass, and with the other improvements in this set for removing alwayslink and not needing custom copts there's much less harm in coalescing a lot of targets within a package

Progress on #3843 (and probably fixes several issues I've lost track of).

To reviewer: I'm not particularly concerned with nits in this - it's overall a significant improvement (and I tried to fix more typos than I introduced!) - it'll be much better to land and refine.

benvanik committed:
Replacing use of PROJECT_SOURCE_DIR/PROJECT_BINARY_DIR.
Those use the previously defined project() and that will be wrong if
IREE is included under another project.

benvanik committed:
Simplifying VMA build integration.

benvanik committed:
Sorting copts/linkopts so we can override things.

benvanik committed:
Fixing cmake style issues - prefer `if(` not `if (` please.

benvanik committed:
Pass android libraries as actual linkopts.

benvanik committed:
Documenting iree_copts with a nice big warning.

benvanik committed:
Centralizing -ldl/-lpthread linkopts (as they were in bazel already).

benvanik committed:
Prevent bazel-to-cmake from running on iree/base/CMakeLists.txt for now.

benvanik committed:
Removing legacy repo_utils.bzl.

benvanik committed:
Cleaning up MSVC warnings and syncing with bazel warnings.

benvanik committed:
Refining MSVC warning set to the minimum required and documenting.

benvanik committed:
Fixing various type warnings.
We don't today have these warnings enabled in our code but it'd be nice
to remove them from the list at some point. This starts us down that path
by hopefully making it easier for code authors to copy/paste or at least
notice the right patterns.

benvanik committed:
Removing unneeded flatbuffers copts.

benvanik committed:
Removing unused iree::schemas::reflection_data.

benvanik committed:
Adding include for flatcc to flat_c_libraries.

benvanik committed:
Adding flatcc json support and making iree-dump-module use it.

benvanik committed:
Removing unused tag field in metal/spirv.

benvanik committed:
Replacing compiler use of HAL executable flatbuffers->flatc.

benvanik committed:
Replacing runtime use of VMLA executable flatbuffers->flatc.

benvanik committed:
Replacing runtime use of dylib executable flatbuffers->flatcc.

benvanik committed:
Replacing runtime use of llvmjit executable flatbuffers->flatcc.

benvanik committed:
Replacing runtime use of spirv executable flatbuffers->flatcc.

benvanik committed:
Replacing runtime use of metal executable flatbuffers->flatcc.
Maybe it works? Who knows! Hooray having no CI coverage.

benvanik committed:
Replacing compiler use of VM bytecode module def flatbuffers->flatcc.

benvanik committed:
Removing all build config/utils related to flatbuffers.

benvanik committed:
byte->ubyte in flatbuffer defs.

benvanik committed:
Goodbye flatbuffers (well, the C++ ones anyway).

benvanik committed:
Switching from -DVK_NO_PROTOTYPES to iree/hal/vulkan/vulkan_headers.h.

benvanik committed:
Adding iree/base/flatcc.h to make flatcc easier to include.

benvanik committed:
Remove unused args from flatbuffers_c_library macro.

benvanik committed:
Adding IREE_TARGET_GUI_LINKOPTS to remove custom linkopts use.

benvanik committed:
Simplifying iree/hal/vmla/ build config.

benvanik committed:
Simplifying iree/hal/metal/ build config.

benvanik committed:
Simplifying iree/hal/llvmjit/ build config.

benvanik committed:
Simplifying iree/hal/vulkan/ build config.

benvanik committed:
Simplifying iree/hal/ build config.

benvanik committed:
Simplifying iree/hal/dylib/ build config.

benvanik committed:
Removing unused PLATFORM_VULKAN_DEPS.
It may be needed internally but it shouldn't clutter up the OSS repo.

benvanik committed:
Replace absl::exchange with iree::exchange to reduce absl/utility dep.

benvanik committed:
Removing unused absl/types/variant.h include.

benvanik committed:
Replacing uses of some absl utilities now in C++14.

benvanik committed:
Rearranging iree/vm/ to reduce a public + cc target.

benvanik committed:
TODOs on future library layout changes.

benvanik committed:
Removing emitc usage of IREE_COMMON_INCLUDE_DIRS for now.

benvanik committed:
Removing IREE_COMMON_INCLUDE_DIRS mhlo pollution.

benvanik committed:
Removing IREE_COMMON_INCLUDE_DIRS and uses by LLVM/MLIR.

benvanik committed:
Changing iree-tblgen to not require alwayslink.

benvanik committed:
Removing ALWAYSLINK support from external_cc_library.

benvanik committed:
Changing to an explicit driver registration mechanism.
This is required for half a dozen reasons but it also enables tons of
cleanup and simplification of the build config.

benvanik committed:
Removing iree/base/initializer.h.

benvanik committed:
Removing ALWAYSLINK support from cmake.

benvanik committed:
Removing iree_api_init from the API.

benvanik committed:
Changing iree::InitializeEnvironment to iree_flags_parse.
Preparation for #3814.

benvanik committed:
Fixing bad type in hal/api.h (been there for ages!).

ScottTodd reviewed (commented):
Looks great! I see your

> To reviewer: I'm not particularly concerned with nits in this - it's overall a significant improvement (and I tried to fix more typos than I introduced!) - it'll be much better to land and refine.

comment, so most of these comments can be ignored or addressed later. It's easier to collect them commit-by-commit here than it is scanning through the whole repo for things to tidy up though.

benvanik committed:
Removing build-config setting of _GNU_SOURCE.

# Conflicts:
#	iree/base/threading_pthreads.c
#	iree/base/threading_win32.c

benvanik committed:
Rearranging iree/base build rules.
By moving the dynamic_library_test out of the main file we can now use
bazel-to-cmake here. This also replaces the iree_cc_binary aliasing that
was happening in bazel to get -ldl - there's a much better way and I was
able to get rid of our custom shim rules and turn all iree_cc_binaries
transitively dependent on iree/base:dynamic_library or :threading to
standard cc_binaries.

benvanik committed:
Removing _GNU_SOURCE copt for wait_handle.

benvanik committed:
Removing the broken forward declarations entirely from some codegen code.
https://google.github.io/styleguide/cppguide.html#Forward_Declarations
Forward declarations are almost never a good idea, and definitely never a
good idea when crossing project boundaries (like iree->mlir/llvm).

benvanik committed:
Speculative removing nowindows tags (#3615).
If there's something that still doesn't work on bazel it's something
we'll need to fix with bazel stuff.

benvanik committed:
Overriding the default MLIR -> LLVM module name.

benvanik committed:
Skipping dylib driver in simple_embedding_test as a hack.

ScottTodd committed:
Add package to iree/base/testing/BUILD.

ScottTodd committed:
Disable layering_check in iree/hal/drivers/BUILD.

ScottTodd committed:
Drop IREE_DRIVER_MODULES and iree/base:initializer from ModelBuilder.

ScottTodd committed:
Fix warning flag on Windows and HAL CTS driver registration. (#3911)

* Wrap warning flag in select() for Windows compatibility.
* Register HAL drivers *before* instantiating HAL CTS test suites.

ScottTodd committed:
Only register VMLA driver in bindings/java/.

This avoids an unconditional dep on all drivers (mainly to avoid LLVMJIT which can't build on Android).

GMNGeoffrey reviewed (approved):
I'm not thrilled with the size of this PR, but overall the simplifications look good

benvanik committed:
Simplify dylib driver switch.

Co-authored-by: Geoffrey Martin-Noble <gcmn@google.com>

benvanik committed:
Reverting flatcc to use our own cmake file for cross-compilation.
